### PR TITLE
adds support for meta (alt) keyboard shortcuts

### PR DESCRIPTION
**Pull request by joemfb**
Also adds implementations of meta-. and meta-backspace;  they're working (ie, tested with other keys), but commented out. Wiring them up will take changes to `dill-belt`, and the code-paths that construct instances of it. I tried changing `mar/dill/belt/hoon`, but it didn't seem to make a difference ...

Questions:

- where should the utility functions go (and how should they be named)?
- is there a better pattern to use in `++alnum`?
- what do I need to change to get meta+ backspace or `.` combinations into the `++ta-met` codepath?

### CODE REVIEW

**Comment by ohAitch** [comment]
Ooh! Any chance of alt-arrows, if alt-backspace can work?

Dill-belt is in arvo, specifically zuse and dill(and perhaps the c code
backing dill). I'm not sure why only letters are coming through.

**Comment by ohAitch** [comment]
Ooh! Any chance of alt-arrows, if alt-backspace can work?

Dill-belt is in arvo, specifically zuse and dill(and perhaps the c
interface driving it). Not sure why only letters would be coming through.

**Comment by joemfb** [comment]
Woooah, I just tried and they already work! Something's already normalizing the input ... That's convenient, and maybe a clue for meta-. and meta-backspace

**Comment by ohAitch** [comment]
I've seen `%met` stub messages when I reach for Alt keys, yeah

**Comment by cgyarvin** [comment]
This is because you're resurrecting a feature that used to exist :-)

**Comment by joemfb** [comment]
Well, *requesting* said resurrection ...

It looks like the input normalization is in `mar/dill/belt/hoon`: https://github.com/urbit/arvo/blob/master/mar/dill/belt.hoon#L39... I had tried changing that file, adding printfs and changing behavior, but none of those changes seemed to take. I've seen seen the same behavior with other files before (`app/talk/hoon`)... I shouldn't need to `|reset` or `|reload` specific vanes for any files in `app/`, `lib/`, `mar/`, etc., right?

**Comment by joemfb** [comment]
It actually looks like some characters are rewritten in `term.c`: https://github.com/urbit/urbit/blob/master/vere/term.c#L818

**Comment by joemfb** [comment]
meta-. and meta-backspace work now (in conjunction with urbit/urbit#758) ...

It looks like bash also implements meta t,u,l,c, and r (http://ss64.com/bash/syntax-keyboard.html). Interested in any or all of those?

**Comment by cgyarvin** [comment]
(Claps.). Sure!

Any general thoughts on the application system as you've learned it?  Questions?

**Comment by joemfb** [comment]
(Commenting here -- still gun-shy about replying to the email thread after #191. And I have a bad habit of editing my github comments after posting, which, I just realized doesn't play well with the email notifications...)

Still coming to grips with the overall application system; I'm really just nibbling around the edges with these PRs. And I have fractals of questions, as usual ;)

As for `:dojo`, it's pretty cool to hotpatch low-level components so easily. It's not yet easy for me to follow the control flow by reading, but I've just added printfs liberally. (It'd help if there was a way to output a stack trace without crashing... And an analogue for inspecting wires, when I get a better handle on moves)

For this PR: where should I put "utility" gates I've added? And is there some Hoonian alternative to regexps, or do I have to just assert the codepoint ranges as I've done?

**Comment by cgyarvin** [comment]
Printfs are a great way to debug Hoon.  We love printfs (~&).

(Sorry about the **** stack traces (^C and stack overflows crash instead of
producing a good trace).  That should be a higher priority bug.  It's
worked for most of Urbit's history, we just let it break while we, you
know, made videos and stuff.  Don't miss the stacktrace printf rune ~|,
though.  Everyone has printfs, but not everyone has stacktrace printfs.)

There's nothing wrong with enumerating the codepoint ranges.  It's not like
those ranges are going to change.

If a core could be used by arbitrary userland code, but isn't likely to be
needed in the kernel, it goes in zuse.hoon.  Zuse has an unfortunate
resemblance to a garbage dump, which should be corrected at some point.

Turning on |verb will show you all vane-moves that are made.  It's a toggle
command.

We had a jetted regexp library, but it was kind of poor quality and so we
threw it out.  Would be nice to have another.

**Comment by joemfb** [comment]
I had missed `~|`, that's perfect!

**Comment by joemfb** [comment]
I've added meta-c,l,t, and u. I need to spend more time with the history implementation to figure out meta-r.

One problem: certain combinations are supposed to be cumulative in the kill-ring -- meta-backspace, meta-backspace, ctrl-y should yank both kills, while ctrl-k, meta-backspace should just yank the latter. I'm not yet sure, but I think the `kil` structure will need to be enhanced with some kind of flag to indicate the nature of the previous kill.

Also, this entire PR could probably use a Hoon style review; from faces to whitespace to favoring slag/scag over snag -- I can defend my choices, but that doesn't make them right ...

**Comment by juped** [comment]
I would probably slag and scag too, because they just do what they should, unlike the *stupid* zero-based snag.

**Comment by juped** [comment]
As a point of style, you may be leaning too hard on `=+` to use a top-down "imperative" style rather than an outside-in style, but you may also not be - sometimes you really do need to control the order of computation. Definitely `=|` a mold for things like your loop counters, though.

**Comment by joemfb** [comment]
Thanks! Some of those `=+` twigs are just trying to control the right margin; there's a lot of intermediate gate calls on positional deltas, for instance

File: lib/drum.hoon [hunk 1]
```diff
@@ -698,10 +698,115 @@
...
+            =+  ^=  prod
+            %-  ta-hom
+            :~  %mor
+              (ta-cut p.next (sub q.next p.next))
+              (ta-cat p.next (slag p.prev (scag q.prev buf.say.inp)))
+              (ta-cut p.prev (sub q.prev p.prev))
+              (ta-cat p.prev (slag p.next (scag q.next buf.say.inp)))
+            ==
+            prod(pos.inp q.next)
+            ::
```
Comment (joemfb) [inline lib/drum.hoon]: This is awkward, but lets me order the edits in a straightforward manner. The alternative includes a lot more offset calculation...

File: lib/drum.hoon [hunk 2]
```diff
@@ -698,10 +698,115 @@
...
+            ?:  =(pos.inp (lent buf.say.inp))
+              ta-bel
+            =+  case=?:(?=($u key) cuss cass)
+            =+  sop=(add pos.inp (next-word (slag pos.inp buf.say.inp)))
+            =+  f=(jump-fwrd sop buf.say.inp)
+            %-  ta-hom
+            :~  %mor
+              (ta-cut sop (sub f pos.inp))
+              (ta-cat sop (tuba (trip (case (tufa (slag sop (scag f buf.say.inp)))))))
+            ==
```
Comment (joemfb) [inline lib/drum.hoon]: This seems absurd, but copying the stdlib gates `++cuss` and `++cass` so I can change the sample mold also seems absurd.

File: lib/drum.hoon [hunk 3]
```diff
@@ -698,10 +698,115 @@
...
+      $bac  ?:  =(0 pos.inp)
+              ta-bel
+            =+  b=(jump-bwrd pos.inp buf.say.inp)
+            %-  ta-hom(kil `(slag b (scag pos.inp buf.say.inp)), ris ~)
+            (ta-cut b (sub pos.inp b))
+            ::
+      $b    ?:  =(0 pos.inp)
+              ta-bel
+            +>(pos.inp (jump-bwrd pos.inp buf.say.inp))
+            ::
```
Comment (joemfb) [inline lib/drum.hoon]: Should I call `++ta-hom` with a `$nop` edit, instead of directly change `+>`?

File: lib/drum.hoon [hunk 4]
```diff
@@ -698,10 +698,115 @@
...
+            =+  sop=(add pos.inp (next-word (slag pos.inp buf.say.inp)))
+            =+  ^=  prod
+            %-  ta-hom
+            :~  %mor
+              [%del sop]
+              :+  %ins  sop
+              (turf (cuss [(tuft (snag sop buf.say.inp))]~))
+            ==
+            prod(pos.inp (jump-fwrd sop buf.say.inp))
+            ::
```
Comment (joemfb) [inline lib/drum.hoon]: This is also awkward, but pretty much required... It's an odd shortcut: a single char edit, then a positional jump.

**Comment by juped** [comment]
Yeah. For right margin control, we have the `%` family to call gates without the attack of the parentheses. But clarity is king above all.

**Comment by juped** [comment]
Also, I don't know what you mean by "copying gates to change the sample mold".

**Comment by joemfb** [comment]
`++cuss` and `++cass` would work as-is for changing the case of a `tuba

**Comment by joemfb** [comment]
oops...

but I have to convert to and from a `tape` to be able to call them, do the the type constraints of their "sample molds" (the mold that normalizes the sample?).

**Comment by juped** [comment]
Using those gates is causing you more trouble than it solves, since they take tapes and produce cords for some weird reason. Helper functions are fine (and give you a place to swap in more comprehensive casing functionality, or replace with the name of an eventual zuse casemapping facility), perhaps defined in a `|^`?

**Comment by joemfb** [comment]
I broke down and trimmed trailing whitespace in this fail... I can back it out of that's a bridge too far.

I added utility gates `++lcas` and `++ucas`; they're wet gates taking `(list @)` right now, but I can't fight the urge to do something like this:

    |*  a/?((list @) @)
    ^+  a
    =+  b=|=(a/@ ?.(&((gte a 'A') (lte a 'Z')) a (add 32 a)))
    ?:  ?=(@ a)
      (b a)
    (turn a b)

That's a `nest-fail`, but it seems like a reasonable piece of code ;) I can do subject surgery to get around it, but I don't know how to do that polymorphically with the list

**Comment by juped** [comment]
It doesn't need to be general, and in fact shouldn't, you're just making a helper function. That's why I suggested a `|^` inside the u/l case. Drum isn't the stdlib.

**Comment by juped** [comment]
Also, your sample code is a dangerous fork abuse. You need to tag your unions most of the time...

    > ? =+  a=*?(@ (list @))  ?>(?=(@ a) a)
      ?(@ $~)
    0

**Comment by joemfb** [comment]
Yeah, i'm just complaining ... It seems like something that should be pretty straightforward, but maybe that's just my dynamic lang experience ...

I need to store two bits of state to support properly cumulative yanks: if the last belt was a yank, and, if so, in which direction. (From the middle of a line, ctrl-u,ctrl-k,ctrl-y should leave the line unchanged. That applies to any consecutive series of yanks, but any other input or navigation should break the chain).

That state could be in one face: `(unit ?($l $r)`, but it'd obviously be disproportionately disruptive to modify all the non-yanking belt codepaths to correctly reset that face. It could be handled in `++se-belt`, modifying the subject before applying the switch expression, but that also seems inelegant. Any thoughts on where that should go?

**Comment by joemfb** [comment]
Too the best of my knowledge, this PR now implements all bash meta shortcuts except for `y`, which would require a real kill-ring.

The "cumulative yanks" issue I described above is a pre-existing problem, but acerbated by meta-d and meta-backspace.

**Comment by juped** [comment]
>bash

You mean emacs keys :angry:

**Comment by joemfb** [comment]
"emacs" keys ... meta-r, ctrl-u are bash only, meta-t, meta-. have different behavior

**Comment by juped** [comment]
You're not allowed to add things that are bash-only, since Urbit is not Ubuntu. :) Fortunately, none of those things are bash-only... just normal shell emacs keys. (Except ^U which is in the terminal driver.)

**Comment by joemfb** [comment]
I went ahead and added ctrl-w, since I had all the pieces ... I also made the utility gates more urbity

**Comment by joemfb** [comment]
I implemented cumulative kills by storing the current and previous belts as a pair in `++target`.

**Comment by joemfb** [comment]
Changing that data structure means a `prep-mismatch` on sync:

```
gall: %hood: prep mismatch
[%talk-peer-replaced 2 i=%~.sole t=/]
[%dojo-peer-replaced 2]
activated app base/dojo
activated app base/talk
```

**Comment by juped** [comment]
Do you know how to write a `++prep`?

**Comment by joemfb** [comment]
I do not, but had meant to ask ... That'd need to go in both `:dojo` and `:talk`, right?

**Comment by juped** [comment]
Anything that uses the library - though you'd write the actual updater in the library itself, of course.

**Comment by ohAitch** [comment]
Hm, the only file I see changed is drum, which means the relevant prep is
%hood, or rather a %hood hookup to in drum itself. I think examples of
those exist, but it definitely isn't a standard app one.

**Comment by joemfb** [comment]
Added a kill-ring, 'cause I can't stop myself ...

`++prep` should be a gate that takes the old app-state and produces the new ... so, I should write a gate in drum that can handle that conversion, and then create a `++prep` arm in hood that gets `%drum` out of  `lac` in `hood-0` and calls into drum to upconvert it?

Will that be sufficient? I'm not really clear on the relationship between dojo and talk have with hood ...

**Comment by juped** [comment]
Drum is the userspace interface to terminal stuff, I think. You need a gate that takes the old state type and produces a new one; see talk for a good example that spans multiple state versions.

Let me know when this is done done.

**Comment by joemfb** [comment]
I'm trying to understand the state-versioning in `:talk`, but I'm stuck on the mold "transformations":

    ++  house-3                                         ::
      %+  cork  house-4  |=  house-4                    ::  modern house with
      +<(stories (~(run by stories) story-3))           ::  old stories
    ++  house-4                                         ::
      %+  cork  house-5  |=  house-5                    ::  modern house with
      +<(shells (~(run by shells) shell-4))             ::  no settings
    ++  house-5                                         ::
      $:  stories/(map knot story)                      ::  conversations
          general/(set bone)                            ::  meta-subscribe
          outbox/(pair @ud (map @ud thought))           ::  urbit outbox
          folks/(map ship human)                        ::  human identities
          shells/(map bone shell)                       ::  interaction state
          log/(map knot @ud)                            ::  logged to clay
          nik/(map (set partner) char)                  ::  bound station glyphs
          nak/(jug char (set partner))                  ::  station glyph lookup
      ==

`cork` is compose-forward, so `++house-3` and `++house-4` call the mold against their sample, then call the gate against the molded sample... That seems backwards, if the sample is an old version of the state, it first molds it against a newer version, than transforms it in some way? Won't the portions of the newer-version mold that *don't* match the sample just bunt?

The other thing that has me stuck is this pattern:

    ++  story-3  (cork story |=(story +<(|10 &11.+<)))

What do `|10` and `&11.+<` mean? (I know `+<` is `+6`, the gate sample)

**Comment by ohAitch** [comment]
The old ones are improper molds, which only work as types. Really they
should probably be
    $_  %*  .  *new-mold
but in some cases you want the new bunt available.

For story-3, +<(|10 &11) is "story without the settings field", which is at
|11 in the new mold. Really the correct solution here is probably adding
new fields near the top, and having the step-down historical types be
consistenty |1.new |2.new etc. instead of new(end-of-new
old-last-field.new).

**Comment by joemfb** [comment]
I can't find any docs for `|wing` or `&wing`; they appear to be yet another tree addressing syntax:

```
> &1:[1 2]
1
> |1:[1 2]
2
> &2:[[1 2] [3 4]]
3
> |2:[[1 2] [3 4]]
4
```

`&n` looks like "head of the cell at *n*" -- `|n` looks like "tail of the cell at *n*". Is that correct?

**Comment by ohAitch** [comment]
Entirely. +>+>+>- and +>+>+>+ respectively, where the total length / number
of derefs is the n in |n or &n.

**Comment by ohAitch** [comment]
I suppose -.+.+.+ and +.+.+.+ might be clearer here, for explaining
=(&4:"abcdef" 'd'), =(|4:"abcdef" "ef")

**Comment by juped** [comment]
In other words, you're not right at all.

    > &0:[1 2 3 4 5 6 7 8 9]
    error
    > |0:[1 2 3 4 5 6 7 8 9]
    [1 2 3 4 5 6 7 8 9]
    > &1:[1 2 3 4 5 6 7 8 9]
    1
    > |1:[1 2 3 4 5 6 7 8 9]
    [2 3 4 5 6 7 8 9]
    > &2:[1 2 3 4 5 6 7 8 9]
    2
    > |2:[1 2 3 4 5 6 7 8 9]
    [3 4 5 6 7 8 9]

and so on

**Comment by ohAitch** [comment]
Idk about not right /at all/: it's translating the tuple meaning of "at
index n" into an actual nock axis, so &2 becomes +<. The examples work
afaict.

**Comment by juped** [comment]
As for the other thing, as the comment says, it's a modern house, but with the old stories, by molding the stories with an old story type. This is a weird intermediate state.

`+<` is axis 6, one of the children of axis 3.

**Comment by joemfb** [comment]
Thanks for the `|n` / `&n` explanations!

I've committed a first attempt in 9713c19... It's kind of a mess, and I'm having trouble testing it. Should I mount a %sandbox desk, cp my changes there, then merge back into %home so that both files change in the same transaction?

I've copied the approach in `ohAitch/base:womb`, except if left full molds for the old state. I'm having trouble writing a gate that transforms some part of a mold inside of a collection (`++source` and `++target` are the molds i've updated).

I assume those changes in `ohAitch/base:womb` will eventually make it back into `urbit/arvo:master` -- how should overlapping state changes be handled? Should I rebase onto that branch?

Also, in `++drum-port` I have to call the molds specifically, I'm not able to cast them. I can't figure out why -- the structures seem to match ...

Sorry for all the questions and confusion :|. This state upgrade is conceptually simple, but I'm having a hell-of-a-time with it.

**Comment by joemfb** [comment]
Sorry for the random pile of questions -- that's kind of a pain ... Here's some more ;)

I have it working now -- history is preserved for both :dojo and :talk. To test and refine it, I merged from a %sandbox desk so that I could apply all the changes in this PR in one transaction. I also found that it was easier to add new faces to the head of the mold, where possible.

I think my changes need to be refactored; I've included the full molds for all state versions, so it's very verbose.

And my question about `womb` still stand: if those changes are going to be merged in the future, should I adapt this branch to include the state-structure changes from that one?

**Comment by joemfb** [comment]
I figured out `++target-0`; any tips for `++source-0` or `++drum-pith-0`?

**Comment by ohAitch** [comment]
Were you looking at the drum-pith-0 on the womb branch? (That'll be a fun
merge)

**Comment by joemfb** [comment]
Yeah, that's what my `womb` questions were in relation to ... Seems like it's inviting disaster to not serialize these changes. (My last question is just about refactoring the molds)

Y'all are already running `womb` on multiple piers, right? If so, it'd make sense for my state changes to layer on top of those. As much as I'm impatient to get this merged, it's much less important in the scheme of things ...

**Comment by ohAitch** [comment]
@juped @cgyarvin, thoughts on merging womb? The relevant drum changes
should definitely make it into the mainline, though womb itself may make
sense as a separate repo. (At least the implementation, a stub would be
needed in hood similar to the current check in of %jael).

**Comment by cgyarvin** [comment]
Super happy to merge all this stuff.

The great renaming breach must also include a general purge of state adapters...

**Comment by juped** [comment]
Womb should have been a separate repo to begin with, though I understand why it ended up not being so. You should not combine separate work in the history; that's really not helpful to the history. Anton has some thoughts on the thorny issues that *do* arise when parallel new versions of a state arise (thorny only because you can't automerge ++prep; we are spoiled by textual three-way merge), but that doesn't quite apply yet. Those questions still need sensible answers.

Though it's off-topic for this thread, it might be necessary for the eventual integration: why is womb in hood? and why *wouldn't* womb be an eventual part of the base system, as there will be 2^16 ticketing authorities in the long term?

**Comment by ohAitch** [comment]
Womb is in hood because this is the canonical workaround for gall app
autostart being broken⌥⌫ intentionally disabled; the main argument for
keeping it in the base system is in fact "we're planning to have all the
newfound star owners use it to distribute planets".

Parallel versions arose due to a feature womb /depends/ on, but which
probably should have already been merged separately(except that it's a
hack, but we don't really have a non-hack solution for |quit and |put).
What is needed here long-term is autogenerated upgrade code for "bag of
fixed-name elements" versioned nouns, perhaps using similar treap
collapsing as cores. Defaulted bunt/drop + also
autogenerated container-walking would probably handle 90% of state
upgrades, and merge cleanly as long as names remained unique.

**Comment by cgyarvin** [comment]
I would put womb in base and make it its own app.

**Comment by cgyarvin** [comment]
Why does womb need to be autostarted?

We could make state adapters a fair bit more convenient in a number of ways.  But we're trying to keep our features frozen...

**Comment by ohAitch** [comment]
Long-term as in "when hoon is significantly unfrozen", yup. This hopefully
isn't happening at /least/ until we get back to 2013-2014 stability levels.

What was the case again for taking out auto-start, at least for local
pokes/peers? My basic reasoning here is "don't force 2^16 users to take an
extra step, unless there's a concrete scenario where the step requires user
authorization", and good defaults over interaction in general.

**Comment by juped** [comment]
Cool. Womb for base 2016. This thread can return to topic now. @joemfb, just do your thing in isolation; if you had to know that parallel branches exist, that would be a major process failure. Let us know when your pull request is a pull request.

**Comment by ohAitch** [comment]
+1

**Comment by cgyarvin** [comment]
We took it out because we didn't understand it.  But please don't make design mistakes for a little bit of convenience.

**Comment by joemfb** [comment]
One last question: should arvo be "safe" on a per-commit basis? That is should the state changes and adapters be included together in a single commit?

**Comment by juped** [comment]
Yeah. Though people skip this all the time. But that's a bad idea.

**Comment by joemfb** [comment]
Ok, that's done. I squashed the last 5 commits.

I've noticed an odd behavior when transitioning state from new back to old: `:talk` dumps it's backlog to the console. I'm not sure what's causing that -- I don't really understand :talk too well yet. I have confirmed that history is preserved when transitioning state from old to new.

Outside of that, I think this is done.

**Comment by juped** [comment]
Talk dumps when it's unlinked and relinked, I hear. It probably shouldn't.

Thanks!

**Comment by joemfb** [comment]
Let me know if you'd like this rebased or adjusted further

**Comment by cgyarvin** [comment]
I continue to be slightly amazed that Joe's gone from zero to writing state
adapters in a month.

Must... resolve... docs PRs...

**Pull request merged**